File: residential_controller.js
class Battery {
  constructor(nb_Floors, nb_elevators) {
    this.nb_Floors = nb_Floors;
    this.nb_elevators = nb_elevators;
    this.column = new Column(nb_floors, nb_elevators);
  }
  // take request of elevator with number of request floor and direction
  Request_Elevator(requested_nb_floor, direction) {
    console.log("Elevator requested at floor number", requested_floor_number);
    sleep(2000);
    console.log(
      "Elevator",
      this.column.elevator_list[0].elevator_number,
      "is currently at floor number",
      this.column.elevator_list[0].current_floor
    );
    sleep(2000);
    console.log(
      "Elevator",
      this.column.elevator_list[1].elevator_number,
      "is currently at floor number",
      this.column.elevator_list[1].current_floor
    );
    sleep(2000);
    // find elevator with the nearest one
    var nearest_elevator = this.FindElevator(requested_nb_floor, direction);
    console.log("Returning Elevator", nearest_elevator.elevator_number);
    sleep(2000);
    console.log(
      "Elevator",
      nearest_elevator.elevator_number,
      "is going",
      direction
    );
    nearest_elevator.addFloorToList(requested_nb_floor);
    nearest_elevator.move_next();
  }
  // take a request of number of floor with activate button
  Request_floor(elevator, requested_nb_floor) {
    console.log("Floor", requested_nb_floor, "is requested");
    elevator.activate_inside_button(requested_nb_floor);
    elevator.addFloorToList(requested_nb_floor);
    elevator.moveNext();
  }
  // method too find the best elevator
  FindElevator(requested_nb_floor, direction) {
    let best_difference = 10;
    let best_elevator = null;
    for (var i = 0; i < this.column.elevator_list.length; i++) {
      var difference_floor = Math.abs(
        requested_nb_floor - this.column.elevatorList[i].current_floor
      );
      if (difference_floor < best_difference) {
        best_difference = difference_floor;
        best_elevator = i;
      }
    }
    return this.column.elevatorList[bestElevator];
  }
}
//========================================================== COLUMN =======================================================

class Column {
  constructor(nbFloors, nbElevators) {
    this.nb_floors = nbFloors;
    this.nb_elevators = nbElevators;
    this.elevator_list = [];
    this.call_elevator_list = [];
    this.outside_call_buttons_list = [];
    this.create_elevators();
    this.create_outside_buttons();
  }
  // create elevator and put it in the elevator list
  create_elevators() {
    for (let i = 0; i < this.nb_elevators; i++) {
      let elevator = new Elevator(i + 1, this.nb_floors);
      this.elevator_list.push(elevator);
    }
  }
  // create outside button  and add it too call_button_list
  create_outside_buttons() {
    for (let i = 0; i < this.nb_floors; i++) {
      if (i != this.nb_floors - 1) {
        this.outside_call_buttons_list.push(new Outside_button(i, "Up"));
      }
      if (i != 0) {
        this.outside_call_buttons_list.push(new Outside_button(i, "Down"));
      }
    }
  }
  // it call elevator list with number of requested
  //and send you elevator 
  addToCallElevatorList(requested_nb_floor) {
    this.call_elevator_list.push(requested_nb_floor);
    if (this.direction == "Up") {
      this.call_elevator_list.sort();
    } else if (this.direction == "Down") {
      this.call_elevator_list.sort().reverse();
    }
  }
}
class Outside_button {
  constructor(requFloor, direction) {
    this.requFloor = requFloor;
    this.direction = direction;
    this.activateButton = false;
  }
}
//========================================================== ELEVATOR ==========================================================================
// class elevator for all initial parameter 
class Elevator {
  constructor(elevator_number, nb_floors) {
    this.elevator_number = elevator_number;
    this.nb_floors = nbFloors;
    this.direction = "Stop";
    this.status = "Idle";
    this.request_floor_list = [];
    this.inside_request_button_list = [];
    this.current_floor = 0;
    this.create_inside_buttons();
  }
// create inside button and add it too a list 
  create_inside_buttons() {
    for (var i = 0; i < this.nbFloors; i++) {
      this.inside_request_button_list.push(new Inside_button(i));
    }
  }
  // create method for send the best elevator
  moveNext() {
    let requested_nb_floor = this.request_floor_list.shift();
    if (this.current_floor === requested_nb_floor) {
      this.OpenDoor();
    } else {
      while (this.current_floor != requested_nb_floor) {
        if (this.current_floor > requested_nb_floor) {
          this.moveDown();
          console.log(this.current_floor, "Going Down");
          sleep(2000);
        } else if (this.current_floor < requested_nb_floor) {
          this.moveUp();
          console.log(this.current_floor, "Going up");
          sleep(2000);
        }
      }
    }
    console.log(
      "Elevator",
      this.elevator_number,
      "arrived at floor number",
      this.current_floor
    );
    // open door and closed it aftert 2 sec
    this.OpenDoor();
    sleep(2000);
    this.CloseDoor();
  }
  moveDown() {
    this.direction = "Down";
    this.status = "Moving";
    this.current_floor = this.current_floor - 1;
    sleep(2000);
  }
  moveUp() {
    this.direction = "Up";
    this.status = "Moving";
    this.current_floor = this.current_floor + 1;
    sleep(2000);
  }
  // create request floor list 
  addFloorToList(requestedFloorNumber) {
    this.request_floor_list.push(requested_nb_floor);
    if (this.direction == "Up") {
      this.request_floor_list.sort();
    } else if (this.direction == "Down") {
      this.request_floor_list.sort().reverse();
    }
  }
  activateInsideButton(requested_nb_floor) {
    console.log(
      "In Elevator",
      this.elevator_number,
      ", which is at floor",
      this.current_floor,
      ",",
      "floor number",
      requested_nb_floor,
      "is requested"
    );
    if (this.requFloor == this.request_floor_list) {
      this.activate_inside_button = false;
    }
    if (this.requFloor < this.request_floor_list) {
      this.moveUp();
    } else if (this.requFloor > this.request_floor_list) {
      this.moveDown();
    }
  }
  OpenDoor() {
    console.log("Opening doors at floor number", this.current_floor);
    //this.status = "openDoor";
  }
  CloseDoor() {
    console.log("Closing doors");
    //this.status = "closeDoor";
  }
}
//=================================================================== REQUEST BUTTON ==================================================================
class inside_button {
  constructor(floor) {
    this.floor = floor;
    this.status = "desactivated";
  }
}
// function for add time 
function sleep(milliseconds) {
  var start = new Date().getTime();
  for (var i = 0; i < 1e7; i++) {
    if (new Date().getTime() - start > milliseconds) {
      break;
    }
  }
}
//=====================================================================================================================================================
const battery = new Battery(10, 2);
////==================================TEST 1========================////
battery.column.elevator_list[0].current_floor = 1;
battery.column.elevator_list[1].current_floor = 1;
battery.Request_elevator(4, "UP");
battery.Request_floor(battery.column.elevator_list[1], 7);
////=========================TEST 2====================================////
// battery.column.elevator_list[0].current_floor = 10;
// battery.column.elevator_list[1].current_floor = 3;
// battery.Request_elevator(1, "Up");
// battery.Request_floor(battery.column.elevator_list[], 6);
// battery.Request_elevator(3, "Up");
// battery.Request_floor(battery.column.elevator_list[], 5);
// battery.Request_elevator(9, "Down");
// battery.Request_floor(battery.column.elevator_list[], 2);
////========================TEST 3========================================////
// battery.column.elevator_list[0].current_floor = 10;
// battery.column.elevator_list[1].current_floor = 3;
// battery.Request_elevator(10, "Down");
// battery.Request_floor(battery.column.elevator_list[], 3);
// battery.Request_elevator(3, "Down");
// battery.Request_floor(battery.column.elevator_list[], 2);

File: residential_controller.py
import time
import math


# this  is the initial system for start the system for columns and elevator when is status is operational all system turn it on
print("Initial Battery Is Now Operational")
class Battery(object):
    def __init__(self, nb_column, nb_elevator):
        self.status = "OPERATIONAL"
        self.nb_column = nb_column
        self.column_list = []
        self.generate_columns(nb_column, nb_elevator)


  
                           
#generate_column and add in a column list
    def generate_columns(self, nb_column, nb_elevator):
        for x in range(self.nb_column):
            self.column_list.append(Column("Column"+str(x+1), 2, 10))

print("Starting Column System: ")
class Column(object):
    def __init__(self, column_name, nb_elevator, nb_floor):
        self.name = column_name
        self.status = "Operational"
        self.nb_elevator = nb_elevator
        self.nb_floor = nb_floor
        self.elevator_list = []
        self.generate_elevator()
        self.call_buttons = []
        self.generate_button()
        self.column_list = []
        print(column_name, "In Direction")
    


    def find_call_button(self, requested_floor, direction):
        best_elevator = None
        for button in self.elevator_list:
            if button.requested_floor == requested_floor and button.direction == direction:
                best_elevator = button
                print("FOUND BUTTON HAVE BEEN PRESSEND"+str(requested_floor))
                break
        return best_elevator


    def RequestElevator(self, requested_floor, direction):
        best_elevator = None
        nearest_elevator = self.nearestElevator(requested_floor)
        print("Nearest Elevator Is: " + str(nearest_elevator.name))
        for elevator in self.elevator_list:

            if requested_floor == elevator.current_floor:
                best_elevator = elevator
                print("The Best Elevator Is Choice:")
                # elevator.elevator_Doors()
              
            elif elevator == nearest_elevator:
                best_elevator = elevator 
                print("Second Nearest Elevator is Choice:")
                # elevator.elevator_Doors()
        
            elif elevator.status == "Stopped" and elevator.current_floor == requested_floor:
                best_elevator = elevator
                print("Third Nearest Elevator Is Choice:")
                # elevator.elevator_Doors()

            elif elevator.status == "Idle" and elevator.current_floor == requested_floor:
                best_elevator = elevator
                print("fourth Nearest Elevator Is Choice:")
                # elevator.elevator_Doors()

            elif elevator.status == "STOPPED" and direction == "GoingUp" and elevator.current_floor == requested_floor:
                best_elevator = elevator
                print("Fift Nearest Elevator Is Choice:")
                # elevator.elevator_Doors()

            elif elevator.status == "STOPPED" and direction == "GoingDown" and elevator.current_floor == requested_floor:
               best_elevator = elevator
               print("sixth Nearest Elevator Is Choice:")
        print("The Nearest Elevator: " + str(best_elevator.name)+str(requested_floor))
        best_elevator.floor_list.append(requested_floor)
        return best_elevator
    print("CalCulating The Nearest Elevator: ")
                
        
# reference_gap = difference between reference gap and gap between 
# each elevator Reference_gap is Elevator is looking for a distance then it goes around and finds the nearest   
    def nearestElevator(self, requested_floor):
        reference_gap = self.elevator_list[0].current_floor - requested_floor
        for elevator in self.elevator_list:
            gap = elevator.current_floor - requested_floor
            if gap <= reference_gap:
                reference_gap = gap
                nearest_elevator = elevator
        return nearest_elevator

# how elevator gonna move with alls parameter and return it 
    def moveElevator(self, elevator, requested_floor, direction, current_floor): 
        while requested_floor > elevator.current_floor:
            elevator.status == "moving"
            if elevator.status == "moving" and elevator.direction == "GoingUp" and direction == elevator.direction:
                if elevator.status == "moving" and elevator.direction == "GoingUp" and direction == elevator.direction:
                    return elevator
    #                 IF Status IS Stopped AND ElevatorDirection IS GoingUp AND Direction IS ElevatorDirection THEN
    #                    RETURN Elevator
        while requested_floor < elevator.current_floor:
            elevator.status == "moving"
            if elevator.status == "moving" and elevator.direction == "GoingDown" and direction == elevator.direction:
                if elevator.status == "STOPPED" and elevator.direction == "GoingDown" and direction == elevator.direction:
                    return elevator
    #                 IF Status IS Stopped AND ElevatorDirection IS GoingDown AND Direction IS ElevatorDirection
            if elevator.current_floor == elevator.best_elevator and elevator.floor_list < elevator.current_floor:
                return elevator
            else:
                return elevator

  
    def request_floor(self, elevator, requested_floor):
        # print activate_floor_request_button(Elevator, requested_floor)
        elevator.floor_list.append(requested_floor)
        
# it generate a request of outside elevator up and down and add it too call_button_list
    def generate_button(self):
        for x in range(self.nb_floor):
            if x!=(self.nb_floor -1):
                self.call_buttons.append(("UP", x+1))
            if x!=0:
                self.call_buttons.append(("DOWN", x+1))

# it generate elevator and add it too elevator list with them attribute
#outside button 
    def generate_elevator(self):
        for x in range(self.nb_elevator):
            self.elevator_list.append(Elevator("elevator"+str(x+1), 0, 0))


    def OperateElevator(self):
        for elevator in self.elevator_list:
            while len(elevator.floor_list) > 0:
                if elevator.current_floor == elevator.floor_list[0]:
                    elevator.status = "Stopped"
                    elevator.floor_list.pop(0)
                    elevator.elevator_doors()
                elif elevator.current_floor > elevator.floor_list[0]:
                    elevator.status = "Moving"
                    elevator.Direction = "GoingDown"
                    elevator.move_down(elevator.floor_list[0])
                    elevator.elevator_doors()
                elif elevator.current_floor < elevator.floor_list[0]:
                    elevator.status = "Moving"
                    elevator.Direction = "GoingUp"
                    elevator.move_up(elevator.floor_list[0])
                    elevator.elevator_doors()
            elevator.status = "Idle"
            elevator.Direction = "GoingNoWhere"
    print("Elevator Is Moving: ")

class Elevator(object):   
    def __init__(self, name, origin, nb_floor):
        self.name = name
        self.status = "idle"
        self.direction = "none"
        self.origin = origin
        self.nb_floor = nb_floor
        self.doors = "close"
        self.call_button = []
        self.floor_list = []
        self.current_floor = 1
        self.request_floor_list = []
        self.inside_light_elevator = "OFF"

          
# inside light with turn "off" and "on" 
    def turn_light(self, status, elevator):
        if elevator.inside_light_elevator == "OFF":
            print("light OFF")
        else:
            print("light ON")
            
            
#  making doors open of elevator with a timer of 5 secondes   and closed it after 5 secondes      
    def elevator_doors(self):
            elevator.doors = "Open"
            time.sleep(3)
            print("open_doors")
            elevator.doors = "Closed"
            time.sleep(5)
            print("elevator_doors" == "Closed")


#  create a sequence that moves Elevator up with the Elevator and Elevator position,
#  then it categorizes all the Elevator and if Elevator are bigger and it increases by 1 each time,
#  and repeats until it finds the good stages
    def move_up(self, floor):
        while self.current_floor < floor:
            self.current_floor +=1
            print(" #" + str(self.name) + " is at floor number: " + str(self.current_floor))

#  create a sequence that moves Elevator down with the Elevator and Elevator position,
#  then it categorizes all the Elevator and if Elevator are smaller and it decreases by 1 each time,

    def move_down(self, floor):
        while self.current_floor > floor:
            self.current_floor -=1
            print("#" + str(self.name) + " is at floor number: " + str(self.current_floor))


#   def __init__(self, name, origin, nb_floor):
# elevatorsss = Elevator("coucou", 0, 10,)
# print("first.elevatorsss.status =" + elevatorsss.status)

# print("Debut test de base remi*****************************")
# column = Column("column_name", 2, 10)
# print("first column status = " + column.status)

# for elevator in column.elevator_list:
#     print(elevator)
#     print(elevator.name)
# print("FIN test de base remi*****************************")

# for elevator in column.call_buttons:
#     print("call_buttons")


# print("Debut remi test scenario 1 exigence*****************************")
# column_test_scenario_1 = Column("column_name", 2, 10)



# /////////////////// TEST /////////////////

battery1 = Battery(1, 2)

battery1.column_list[0].elevator_list[0].current_floor = 10
battery1.column_list[0].elevator_list[1].current_floor = 9

elevator = battery1.column_list[0].RequestElevator(6, "UP")
battery1.column_list[0].OperateElevator()

battery1.column_list[0].request_floor(elevator, 8)
battery1.column_list[0].OperateElevator()



